src/linker/Linker.Dataflow/CustomAttributeSource.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Mono.Cecil;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace Mono.Linker.Dataflow
+{
+	class CustomAttributeSource
+	{
+		private readonly XmlFlowAnnotationSource[] _sources;
+
+		public CustomAttributeSource (LinkContext _context)
+		{
+			Collection<XmlFlowAnnotationSource> annotationSources = new Collection<XmlFlowAnnotationSource> ();
+			if (_context.AttributeDefinitions != null && _context.AttributeDefinitions.Count > 0) {
+				foreach (string a in _context.AttributeDefinitions) {
+					XmlFlowAnnotationSource xmlAnnotations = new XmlFlowAnnotationSource (_context);
+					xmlAnnotations.ParseXml (a);
+					annotationSources.Add (xmlAnnotations);
+				}
+			}
+			_sources = annotationSources.ToArray ();
+		}
+
+		public IEnumerable<CustomAttribute> GetCustomAttributes (ICustomAttributeProvider provider)
+		{
+			IEnumerable<CustomAttribute> aggregateAttributes = null;
+			foreach (var source in _sources) {
```

Comment:
We should probably optimize for case where there are no XML sources - so wrap the foreach with if _sources.Length > 0.